src/main/java/org/carlspring/cloud/storage/s3fs/S3Factory.java

```diff
@@ -0,0 +1,326 @@
+    }
+
+    private Region getRegion(final Properties props)
+    {
+        if (props.getProperty(REGION) != null)
+        {
+            return Region.of(props.getProperty(REGION));
+        }
+
+        try
+        {
+            return new DefaultAwsRegionProviderChain().getRegion();
+        }
+        catch (final SdkClientException e)
+        {
+            LOGGER.warn("Unable to load region from any of the providers in the chain");
+        }
+
+        return DEFAULT_REGION;
+    }
+
+    protected SdkHttpClient getHttpClient(final Properties props)
+    {
+        final ApacheHttpClient.Builder builder = getApacheHttpClientBuilder(props);
+        final AttributeMap map = getAttributeMap(props);
+        return builder.buildWithDefaults(map);
+    }
+
+    private ApacheHttpClient.Builder getApacheHttpClientBuilder(final Properties props)
+    {
+        final ApacheHttpClient.Builder builder = ApacheHttpClient.builder();
+
+        if (props.getProperty(CONNECTION_TIMEOUT) != null)
+        {
+            final Duration duration = Duration.ofMillis(Long.parseLong(props.getProperty(CONNECTION_TIMEOUT)));
+            builder.connectionTimeout(duration);
+        }
+
+        if (props.getProperty(MAX_CONNECTIONS) != null)
+        {
+            final int maxConnections = Integer.parseInt(props.getProperty(MAX_CONNECTIONS));
+            builder.maxConnections(maxConnections);
+        }
+
+        if (props.getProperty(SOCKET_TIMEOUT) != null)
+        {
+            final Duration duration = Duration.ofMillis(Long.parseLong(props.getProperty(SOCKET_TIMEOUT)));
+            builder.socketTimeout(duration);
+        }
+
+        return builder.proxyConfiguration(getProxyConfiguration(props));
+    }
+
+    private AttributeMap getAttributeMap(Properties props)
+    {
+        final AttributeMap.Builder builder = AttributeMap.builder();
+        if (props.getProperty(SOCKET_TIMEOUT) != null)
+        {
+            final Duration duration = Duration.ofMillis(Long.parseLong(props.getProperty(SOCKET_TIMEOUT)));
+            builder.put(SdkHttpConfigurationOption.READ_TIMEOUT, duration);
+        }
+        return builder.build();
+    }
+
+    /**
+     * should return a new S3Client
+     *
+     * @param builder S3ClientBuilder mandatory.
+     * @return {@link software.amazon.awssdk.services.s3.S3Client}
+     */
+    protected abstract S3Client createS3Client(final S3ClientBuilder builder);
+
+    protected AwsCredentialsProvider getCredentialsProvider(final Properties props)
+    {
+        AwsCredentialsProvider credentialsProvider;
+        if (props.getProperty(ACCESS_KEY) == null && props.getProperty(SECRET_KEY) == null)
+        {
+            credentialsProvider = DefaultCredentialsProvider.create();
+        }
+        else
+        {
+            final AwsCredentials awsCredentials = getAwsCredentials(props);
+            credentialsProvider = StaticCredentialsProvider.create(awsCredentials);
```

Comment:
Do you have code examples, or links to docs where we could get a better understanding of how to improve this?

I believe it would be best to handle this in a follow-up task as well.